ext/libstrawberry-tagreader/gmereader.cpp

```diff
@@ -0,0 +1,284 @@
+
+  /* Music Players that support SPC tend to support additional tagging data as
+   * an APETAG entry at the bottom of the file instead of writing into the xid6
+   * tagging space. This is where a lot of the extra data for a file is stored,
+   * such as genre or replaygain data.
+   *
+   * This data is currently supported by TagLib, so we will simply use that for
+   * the remaining values. */
+  TagLib::APE::File ape(file_info.filePath().toStdString().data());
+  if (ape.hasAPETag()) {
+    TagLib::Tag* tag = ape.tag();
+    if (!tag) return;
+
+    song_info->set_year(tag->year());
+    song_info->set_track(tag->track());
+    TagReaderGME::TagLib_Decode(tag->artist(), song_info->mutable_artist());
+    TagReaderGME::TagLib_Decode(tag->title(), song_info->mutable_title());
+    TagReaderGME::TagLib_Decode(tag->album(), song_info->mutable_album());
+    TagReaderGME::TagLib_Decode(tag->genre(), song_info->mutable_genre());
+  }
+
+  song_info->set_valid(true);
+  song_info->set_filetype(spb::tagreader::SongMetadata_FileType_SPC);
+}
+
+qint16 GME::SPC::GetNextMemAddressAlign32bit(qint16 input) {
+  return ((input + 0x3) & ~0x3);
+  // Plus 0x3 for rounding up (not down), AND NOT to flatten out on a 32 bit
+  // level.
+}
+
+quint64 GME::SPC::ConvertSPCStringToNum(const QByteArray& arr) {
+  quint64 result = 0;
+  for (auto it = arr.begin(); it != arr.end(); it++) {
+    unsigned int num = *it - '0';
+    if (num > 9) break;
+    result = (result * 10) + num;  // Shift Left and add.
+  }
+  return result;
+}
+
+void GME::VGM::Read(const QFileInfo& file_info,
+                    spb::tagreader::SongMetadata* song_info) {
+  QFile file(file_info.filePath());
+  if (!file.open(QIODevice::ReadOnly)) return;
+
+  qLog(Debug) << "Reading tags from VGM file: " << file_info.fileName();
+
+  if (!file.read(4).startsWith(QString("Vgm ").toLatin1())) return;
+
+  file.seek(GD3_TAG_PTR);
+  QByteArray gd3_head = file.read(4);
+  if (gd3_head.size() < 4) return;
+
+  quint64 pt = (unsigned char)gd3_head[0] | ((unsigned char)gd3_head[1] << 8) |
+               ((unsigned char)gd3_head[2] << 16) |
+               ((unsigned)gd3_head[3] << 24);
+
+  file.seek(SAMPLE_COUNT);
+  QByteArray sample_count_bytes = file.read(4);
+  file.seek(LOOP_SAMPLE_COUNT);
+  QByteArray loop_count_bytes = file.read(4);
+  quint64 length = 0;
+
+  if (!GetPlaybackLength(sample_count_bytes, loop_count_bytes, length)) return;
+
+  file.seek(GD3_TAG_PTR + pt);
+  QByteArray gd3_version = file.read(4);
+
+  file.seek(file.pos() + 4);
+  QByteArray gd3_length_bytes = file.read(4);
+  quint32 gd3_length =
+      (unsigned char)gd3_length_bytes[0] | ((unsigned char)gd3_head[1] << 8) |
+      ((unsigned char)gd3_head[2] << 16) | ((unsigned char)gd3_head[3] << 24);
+
+  QByteArray gd3Data = file.read(gd3_length);
+  QTextStream fileTagStream(gd3Data, QIODevice::ReadOnly);
+  // Stored as 16 bit UTF string, two bytes per letter.
+  fileTagStream.setEncoding(QStringConverter::Utf16);
```

Comment:
This does not compile with Qt 5. Need to use a ifdef here and use `QTextStream::setCodec` with Qt 5.

Reply:
This should be addressed now with a compile time conditional.